File: app/src/main/java/net/emojiparty/android/jishotomo/data/room/Entry.java
package net.emojiparty.android.jishotomo.data.room;

import androidx.annotation.NonNull;
import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.Index;
import androidx.room.PrimaryKey;

import java.util.Date;
import java.util.List;
import net.emojiparty.android.jishotomo.data.SemicolonSplit;

@Entity(tableName = "entries", indices = {@Index("jlpt_level")})
public class Entry {
  @PrimaryKey
  @NonNull
  private int id;

  @ColumnInfo(name = "primary_kanji")
  private String primaryKanji;

  @NonNull
  @ColumnInfo(name = "primary_reading")
  private String primaryReading;

  @ColumnInfo(name = "other_kanji")
  private String otherKanji;

  @ColumnInfo(name = "other_readings")
  private String otherReadings;

  @ColumnInfo(name = "jlpt_level")
  private Integer jlptLevel;

  private Date favorited;

  public int getId() {
    return id;
  }

  public void setId(int id) {
    this.id = id;
  }

  public String getPrimaryKanji() {
    return primaryKanji;
  }

  public void setPrimaryKanji(String primaryKanji) {
    this.primaryKanji = primaryKanji;
  }

  public String getPrimaryReading() {
    return primaryReading;
  }

  public void setPrimaryReading(String primaryReading) {
    this.primaryReading = primaryReading;
  }

  public String getOtherKanji() {
    return otherKanji;
  }

  public void setOtherKanji(String otherKanji) {
    this.otherKanji = otherKanji;
  }

  public String getOtherReadings() {
    return otherReadings;
  }

  public void setOtherReadings(String otherReadings) {
    this.otherReadings = otherReadings;
  }

  public Integer getJlptLevel() {
    return jlptLevel;
  }

  public void setJlptLevel(Integer jlptLevel) {
    this.jlptLevel = jlptLevel;
  }

  public List<String> otherReadingsList() {
    return SemicolonSplit.split(otherReadings);
  }

  public List<String> otherKanjiList() {
    return SemicolonSplit.split(otherKanji);
  }

  public Date getFavorited() {
    return favorited;
  }

  public void setFavorited(Date favorited) {
    this.favorited = favorited;
  }

  public boolean isFavorited() {
    return getFavorited() != null;
  }

  public void toggleFavorited() {
    if (this.isFavorited()) {
      this.setFavorited(null);
    } else {
      this.setFavorited(new Date());
    }
  }
}

File: app/src/main/java/net/emojiparty/android/jishotomo/data/room/Entry.kt
package net.emojiparty.android.jishotomo.data.room

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Index
import androidx.room.PrimaryKey
import java.util.Date

/**
 * An Entry has many Senses, which can be thought of like a definition.
 *
 * Ex. an entry with the kanji 行けない and reading いけない has several meanings:
 * 1. wrong, not good, of no use
 * 2. hopeless, past hope
 * 3. must not do
 */
@Entity(tableName = "entries", indices = [Index("jlpt_level")])
data class Entry(
  @PrimaryKey val id: Int,
  @ColumnInfo(name = "primary_kanji") val primaryKanji: String? = null,
  @ColumnInfo(name = "primary_reading") val primaryReading: String,
  @ColumnInfo(name = "other_kanji") val otherKanji: String? = null,
  @ColumnInfo(name = "other_readings") val otherReadings: String? = null,
  @ColumnInfo(name = "jlpt_level") val jlptLevel: Int? = null,
  var favorited: Date? = null
) {
  fun isFavorited(): Boolean {
    return favorited != null
  }

  fun toggleFavorited() {
    favorited = if (isFavorited()) {
      null
    } else {
      Date()
    }
  }

  fun hasKanji(): Boolean = primaryKanji != null

  fun kanjiOrReading(): String = if (hasKanji()) primaryKanji!! else primaryReading

  fun reading(): String? = if (hasKanji()) primaryReading else null
}
